lib/sqlalchemy/dialects/mysql/mariadbconnector.py

```diff
@@ -0,0 +1,182 @@
+    pass
+
+# MariaDB binary protocol doesn't support XA yet, so we need
+# to rewrite the statement. See https://jira.mariadb.org/browse/MDEV-16708
+
+def check_unsupported_xa(statement, parameter):
+    if parameter is None or parameter.__len__() != 1:
+        return None
+    sql = re.sub(re.compile("/\\*.*?\\*/",re.DOTALL ) ,"" ,statement)
+    words= sql.split(None, 1)
+    if words[0].lower() == "xa":
+        if sql.find(" ?"):
+            replace= "'%s'" % parameter[0]
+            return sql.replace("?", replace)
+
+class MySQLDialect_mariadbconnector(MySQLDialect):
+    is_mariadb = True
+    driver = "mariadbconnector"
+    supports_unicode_statements = True
+    encoding = "utf8mb4"
+    convert_unicode = True
+    supports_sane_rowcount = True
+    supports_sane_multi_rowcount = True
+    supports_native_decimal = True
+    default_paramstyle = "qmark"
+    execution_ctx_cls = MySQLExecutionContext_mariadbconnector
+    statement_compiler = MySQLCompiler_mariadbconnector
+    preparer = MySQLIdentifierPreparer_mariadbconnector
+
+    def __init__(self, server_side_cursors=False, **kwargs):
+        super(MySQLDialect_mariadbconnector, self).__init__(**kwargs)
+        self.server_side_cursors = True
+        self.paramstyle= "qmark"
+        if self.dbapi is not None:
+            assert StrictVersion(self.dbapi.__version__) >= StrictVersion(mariadb_cpy_minimum_version),\
+                "The installed version (%s) of MariaDB Connector/Pyton is not supported."\
+                "Please install MariaDB Connector/Python %s or newer" % (self.dbapi.__version__, mariadb_cpy_minimum_version)
+
+    @classmethod
+    def dbapi(cls):
+        return __import__("mariadb")
+
+    def on_connect(self):
+        super_ = super(MySQLDialect_mariadbconnector, self).on_connect()
+
+        def on_connect(conn):
+            if super_ is not None:
+                super_(conn)
+
+        return on_connect
+
+    def is_disconnect(self, e, connection, cursor):
+        if super(MySQLDialect_mariadbconnector, self).is_disconnect(
+            e, connection, cursor
+        ):
+            return True
+        elif isinstance(e, self.dbapi.Error):
+            str_e = str(e).lower()
+            return (
+                "not connected" in str_e or "isn't valid" in str_e
+            )
+        else:
+            return False
+
+    def do_execute(self, cursor, statement, parameters, context=None):
+        xa= check_unsupported_xa(statement, parameters)
+        if xa:
+            cursor.execute(xa, buffered=True)
+        else:
+            cursor.execute(statement, parameters, buffered=True)
+
+    def do_executemany(self, cursor, statement, parameters, context=None):
+        cursor.executemany(statement, parameters)
+
+    def create_connect_args(self, url):
+        opts = url.translate_connect_args()
+
+        int_params = ["connect_timeout", "read_timeout", "write_timeout", "client_flag",
+                     "port", "pool_size"]
```

Comment:
**mike bayer** (zzzeek) wrote:

this isn't needed unless I'm missing something

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2166

Reply:
**Gord Thompson** (gordthompson) wrote:

Let's take it out and see if anything goes boom.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2166

Reply:
**mike bayer** (zzzeek) wrote:

ah OK, in this comment I was referring to just the "do_executemany()" part :)   .

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2166

Reply:
**Gord Thompson** (gordthompson) wrote:

Oops. Sorry about that.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2166